### PR TITLE
SALTO-3516: Salesforce better multi currency support

### PR DESCRIPTION
**Pull request by tamtamirr**
Improved Multi Currency Organizations support in Salesforce Adapter.

---

- Noise suppression PR: https://github.com/salto-io/salto_private/pull/5019
- You can see the diff in my workspace after the change [here](https://github.com/salto-io/tamir-sf/compare/f6bb15235d05e64636b506f694040e2c9789bc5f..7620f97d00d500f71205fd6564cdd1adf9b748b9).

Organizations that support multi currencies, have extra field on some of their data (`CurrencyIsoCode`) which specifies the
coin to use for a specific instance. (e.g. Product, Quantity etc..)

Prior to this PR, this instances had it has hidden value. Also, there was no good experience when trying to clone between orgs that have Multi currency enabled to ones with no multi currency, and vice versa.

This PR handles two things:
- Making the value of the `CurrencyIsoCode` on data instances visible.
- Added a change validator that omits changes that have weird side effects or will fail when cloning instances from
 an org with `MultiCurrency` to one without.

---
_Release Notes_: 
Salesforce Adapter:
- Better support in organizations that have Multi Currency enabled. 

---
_User Notifications_: 
Salesforce:
- For organizations with Multi Currency enabled: You may see the field `CurrencyIsoCode` on some of your data instances, that was previously hidden.

### CODE REVIEW

File: packages/salesforce-adapter/src/constants.ts [hunk 1]
```diff
@@ -320,6 +320,9 @@ export const DEFAULT_CUSTOM_OBJECTS_DEFAULT_RETRY_OPTIONS = {
 }
 export const MAX_TYPES_TO_SEPARATE_TO_FILE_PER_FIELD = 20
 
+// Fields
+export const CURRENCY_ISO_CODE = 'CurrencyIsoCode'
```
Comment (yuribro) [inline packages/salesforce-adapter/src/constants.ts:324]: Can you also use this constant in packages/salesforce-adapter/src/filters/currency_iso_code.ts ?

File: packages/salesforce-adapter/src/change_validators/currency_iso_codes.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+  _.isString(instance.value[CURRENCY_ISO_CODE])
+)
+
+const isValidCurrencyIsoCodesInstance = (instance: InstanceElement): instance is CurrencyIsoCodesInstance => {
+  const valueSet = instance.value[FIELD_ANNOTATIONS.VALUE_SET]
+  return isDefined(valueSet) && makeArray(valueSet)
+    .every(entry => _.isString(entry[INSTANCE_FULL_NAME_FIELD]))
+}
+
+const createOrgHasNoMultiCurrencyEnabledError = (
```
Comment (amitgoldarb) [inline packages/salesforce-adapter/src/change_validators/currency_iso_codes.ts:65]: @tomermevorach WDYT?

File: packages/salesforce-adapter/src/change_validators/currency_iso_codes.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+  severity: 'Error',
+})
+
+const createInstanceHasUnsupportedCurrencyError = (
+  instance: InstanceWithCurrencyIsoCode,
+  supportedIsoCodes: string[],
+): ChangeError => ({
+  elemID: instance.elemID,
+  message: 'Instance has unsupported currency',
+  detailedMessage: `The instance ${instance.elemID.getFullName()} has unsupported currency ${instance.value[CURRENCY_ISO_CODE]}. Supported currencies: ${supportedIsoCodes}`,
```
Comment (amitgoldarb) [inline packages/salesforce-adapter/src/change_validators/currency_iso_codes.ts:80]: @tomermevorach WDYT?

File: packages/salesforce-adapter/src/change_validators/currency_iso_codes.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+}
+
+type InstanceWithCurrencyIsoCode = InstanceElement & {
+  value: InstanceElement['value'] & {
+    [CURRENCY_ISO_CODE]: string
+  }
+}
+
+const isInstanceWithCurrencyIsoCode = (instance: InstanceElement): instance is InstanceWithCurrencyIsoCode => (
+  _.isString(instance.value[CURRENCY_ISO_CODE])
```
Comment (amitgoldarb) [inline packages/salesforce-adapter/src/change_validators/currency_iso_codes.ts:56]: Shouldn't this be isDefined?

File: packages/salesforce-adapter/src/change_validators/currency_iso_codes.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+)
+
+const isValidCurrencyIsoCodesInstance = (instance: InstanceElement): instance is CurrencyIsoCodesInstance => {
```
Comment (amitgoldarb) [inline packages/salesforce-adapter/src/change_validators/currency_iso_codes.ts:59]: Nit: I think valid is redundant in a type gaurd
```suggestion
const isCurrencyIsoCodesInstance = (instance: InstanceElement): instance is CurrencyIsoCodesInstance => {
```

File: packages/salesforce-adapter/src/change_validators/currency_iso_codes.ts [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    .filter(isInstanceChange)
+    .filter(isAdditionOrModificationChange)
+    .filter(change => isInstanceWithCurrencyIsoCode(getChangeData(change)))
+    .toArray() as Change<InstanceWithCurrencyIsoCode>[]
+
+  if (_.isEmpty(changesWithCurrencyIsoCode)) {
+    return []
+  }
+
+  const currencyIsoCodesInstance = await elementsSource.get(new ElemID(SALESFORCE, CURRENCY_CODE_TYPE_NAME, 'instance'))
```
Comment (amitgoldarb) [inline packages/salesforce-adapter/src/change_validators/currency_iso_codes.ts:98]: This elemId seems wierd, doesn't the instance have a name? Or if it's a setting/singleton shouldn't it end with _config?

**Review by amitgoldarb — COMMENTED** [review]
Added some comments.
@adamro I think I'd like you're feedback here as well, as this a larger change I think.

File: packages/salesforce-adapter/src/change_validators/currency_iso_codes.ts (hunk 6 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/src/change_validators/currency_iso_codes.ts:98]: It's the correct instance elem ID. I've also tested it :)

File: packages/salesforce-adapter/src/change_validators/currency_iso_codes.ts (hunk 4 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/src/change_validators/currency_iso_codes.ts:56]: isString covers `undefined` as-well and more correct here

File: packages/salesforce-adapter/src/change_validators/currency_iso_codes.ts (hunk 5 above)
Comment (tamtamirr) [reply packages/salesforce-adapter/src/change_validators/currency_iso_codes.ts]: v

File: packages/salesforce-adapter/src/change_validators/currency_iso_codes.ts (hunk 3 above)
Comment (tomermevorach) [reply packages/salesforce-adapter/src/change_validators/currency_iso_codes.ts:80]: message: "Unsupported currency ${currency code}"
detailed: "Please set to one of the supported currencies or enable the currency you need directly in salesforce. Current supported currencies are: ${supported currencies list}"

Comment (tamtamirr) [reply packages/salesforce-adapter/src/change_validators/currency_iso_codes.ts:80]: v

**Review by amitgoldarb — APPROVED** [review]
Thanks for the quick fix!
I think we can also add to this CV a validation that we aren't editing the valueSet of the multi-currency supported by the org.
But I think we can do this in the next PR to get this one in first.

**Pull request merged**